quinn-proto/src/tests/mod.rs

```diff
@@ -2181,6 +2208,388 @@ fn packet_splitting_not_necessary_after_higher_mtu_discovered() {
+        1
+    );
+    assert_eq!(
+        client_stats_after_first_ping.frame_rx.acks - client_stats_after_connect.frame_rx.acks,
+        0
+    );
+
+    // Restore the default MTU and send another ping, which will arrive earlier than the dropped one
+    pair.mtu = default_mtu;
+    pair.client_conn_mut(client_ch).ping();
+    pair.drive_client();
+    pair.drive_server();
+    pair.drive_client();
+
+    // Client sanity check (ping sent, one ACK received)
+    let client_stats_after_second_ping = pair.client_conn_mut(client_ch).stats();
+    assert_eq!(
+        client_stats_after_second_ping.frame_tx.ping - client_stats_after_connect.frame_tx.ping,
+        2
+    );
+    assert_eq!(
+        client_stats_after_second_ping.frame_rx.acks - client_stats_after_connect.frame_rx.acks,
+        1
+    );
+
+    // Server checks (single ping received, ACK sent)
+    let server_stats_after_second_ping = pair.server_conn_mut(server_ch).stats();
+    assert_eq!(
+        server_stats_after_second_ping.frame_rx.ping - server_stats_after_connect.frame_rx.ping,
+        1
+    );
+    assert_eq!(
+        server_stats_after_second_ping.frame_tx.acks - server_stats_after_connect.frame_tx.acks,
+        1
+    );
+}
+
+#[test]
+fn single_ack_eliciting_packet_with_ce_bit_triggers_immediate_ack() {
+    let _guard = subscribe();
+    let mut pair = Pair::default();
+    let (client_ch, _) = pair.connect();
+    pair.drive();
+
+    let stats_after_connect = pair.client_conn_mut(client_ch).stats();
+
+    let start = pair.time;
+
+    pair.client_conn_mut(client_ch).ping();
+
+    pair.congestion_experienced = true;
+    pair.drive_client(); // Send ping
+    pair.congestion_experienced = false;
+
+    pair.drive_server(); // Process ping, send ACK in response to congestion
+    pair.drive_client(); // Process ACK
+
+    // Sanity check: the time hasn't advanced in the meantime)
+    assert_eq!(pair.time, start);
+
+    let stats_after_ping = pair.client_conn_mut(client_ch).stats();
+    assert_eq!(
+        stats_after_ping.frame_tx.ping - stats_after_connect.frame_tx.ping,
+        1
+    );
+    assert_eq!(
+        stats_after_ping.frame_rx.acks - stats_after_connect.frame_rx.acks,
+        1
+    );
+    assert_eq!(
+        stats_after_ping.path.congestion_events - stats_after_connect.path.congestion_events,
+        1
+    );
+}
+
+#[test]
+fn ack_frequency() {
+    let _guard = subscribe();
+
+    // ---
+    // Config and initialization
+    // ---
+
+    let mut client_config = client_config();
+    let mut ack_freq_config = AckFrequencyConfig::default();
+    ack_freq_config.ack_eliciting_threshold(10u32.into());
+    Arc::get_mut(&mut client_config.transport)
+        .unwrap()
+        .ack_frequency_config(Some(ack_freq_config))
+        .mtu_discovery_config(None); // To keep traffic cleaner
+
+    let mut pair = Pair::default();
+    pair.latency = Duration::from_millis(10); // Need latency to avoid an RTT = 0
```

Comment:
Why is RTT=0 a problem here? RTT=0 (to within timer precision) can happen in reality.

Reply:
I don't exactly remember the specific reason to avoid an RTT = 0 here. Intuitively, I'd think having an RTT > 0 makes it more straightforward to check that the RTT timeout is working properly in the different parts of the test.

Regarding a minimum value, we could modify the timer interval to be at least `TIMER_GRANULARITY`, since that is in theory the minimum possible interval and using a lower interval than that would have no effect. Not sure if it's worth it to do this, though, since it should have no observable effect.

Reply:
Fair point. On review, I think this tests needs RTT≠0 just so that we can reason about the progression of time within it.